TRUST : Create Mumor functionality / enhance samba logic / add basic synergy with Uka Totlihn

GodModeXI opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->
- The purpose of this PR:

- Mumor: 
    - Adds functionality to Mumor that falls into behaviors described by her BG wiki profile. 
    - Adds synergy modifiers and listeners to manage them dynamically
    - Adds checks for ecosystems (undead) to adjust Samba usage.
    - Adds checks for the players job to adjust Samba usage.
    - [](https://www.bg-wiki.com/ffxi/BGWiki:Trusts#Mumor)

- Uka Totlihn:
    - Minor clean up. 
    - Adds synergy modifiers and listeners to manage them dynamically
    - Corrects teamwork messages.
    - Rearranges gambit order, 
    - Adds checks for ecosystems (undead) to adjust Samba usage.
    - Adds checks for the players job to adjust Samba usage.
    - [](https://www.bg-wiki.com/ffxi/BGWiki:Trusts#Uka_Totlihn)

## Steps to test these changes

1. Load in updated SQL file for Mumor's WS.
2. Restart server to apply.
3. Load in updated Trust luas into the trust directory.
4. Load into the game and summon both trusts.
5. !checklocalvar mumorSynergy to ensure synergy vars are applying.
6. Enter combat and check modifiers WALTZ_POTENTCY and SAMBA_DURATION buffs.
7. Release and / or !hp 0 trusts and !checklocalvar mumorSynergy once despawned that the vars are deleted.

- Mumor Synergy
![image](https://user-images.githubusercontent.com/99501957/195507202-1fa8eafe-7a3a-4a8d-bc24-b9dda865ac48.png)

- Uka Totlihn Synergy
![image](https://user-images.githubusercontent.com/99501957/195507065-668c7f9d-f5a7-4484-bd00-a3b6b8f25e4a.png)


<!-- Clear and detailed steps to test your changes here -->

GodModeXI committed:
Update mob_skill_lists.sql

GodModeXI committed:
Update mumor.lua

GodModeXI committed:
Update mumor.lua

GodModeXI committed:
Update uka_totlihn.lua

GodModeXI committed:
Update mumor.lua

GodModeXI committed:
Update uka_totlihn.lua

GodModeXI committed:
Update mumor.lua

Triming trailing space

GodModeXI committed:
Update uka_totlihn.lua

trim trailing space

GodModeXI committed:
Update mumor.lua

GodModeXI committed:
Update uka_totlihn.lua

GodModeXI committed:
Update mumor.lua

GodModeXI committed:
Update uka_totlihn.lua

GodModeXI committed:
Update uka_totlihn.lua

Barthandalus commented:
Apologies for the formatting mess. 

Corrections are made with the exception of the shadowing upvalue errors related to the listeners checking Vars. 

Considering closing this PR and issuing a new PR branch to apply with clean history. 

Just need to understand how to fix these errors prior to closing and resubmitting the PR. (Help!)

![image](https://user-images.githubusercontent.com/99501957/195606608-dcbfc79c-683e-4cad-8065-294b4b2c35db.jpeg)

GodModeXI committed:
Update mumor.lua

update listener function to prevent upvalue shadowing

GodModeXI committed:
Update uka_totlihn.lua

update listener function to prevent upvalue shadowing

zach2good commented on scripts/globals/spells/trust/uka_totlihn.lua:
```diff
@@ -5,34 +5,77 @@ require("scripts/globals/trust")
-    mob:addSimpleGambit(ai.t.TARGET, ai.c.NOT_STATUS, xi.effect.FINISHING_MOVE_5, ai.r.JA, ai.s.SPECIFIC, xi.ja.QUICKSTEP, 20)
+    -- Dynamic modifier that checks variable on tick to apply
+    mob:addListener("COMBAT_TICK", "UKA_TOTLIHN_CTICK", function(mobArg)
+        if synergy >= 2 then
```
I don't think this synergy variable is guaranteed to be captured for execution in this listener. You'll need to look it up from scratch entirely within the function

GodModeXI replied:
Move the variable into the listener?

zach2good replied:
Do all of your calculations inside of a listener using only what is passed into the listener. In this case: mobArg

GodModeXI replied:
I will revise this.

zach2good commented on scripts/globals/spells/trust/uka_totlihn.lua:
```diff
@@ -5,34 +5,77 @@ require("scripts/globals/trust")
+        master:setLocalVar("mumorSynergy", synergy - 1)
+    end)
 
+    for i = 1, #jobs1 do
```
Since this is set on spawn, these gambits will remain if those partyembers disappear. You'll want to find a dynamic way to do this

GodModeXI replied:
When I did test this code, I started as a warrior and they used drain samba. I left the trusts active and did !changejob Whm 99 and on the next effect wear they switched to haste samba. I thought that the for i loop would maintain the logic.

zach2good replied:
Ah I read that wrong, if it's based on the master then this is fine 👍

zach2good commented on scripts/globals/spells/trust/mumor.lua:
```diff
@@ -2,19 +2,78 @@
+        master:setLocalVar("mumorSynergy", synergy - 1)
+    end)
+
+    mob:addListener("DESPAWN", "MUMOR_DESPAWN", function(mobArg)
```
It would be more dynamic to make a listener that runs on every tick and counts other members thay provide synergy, and then changes mods.

If gambits change based on this, you'll need to adjust the gambit container to support adding and removing them at runtime

GodModeXI replied:
Currently most of the synergy I see is adding and deleting modifiers, though some do change behaviors. I guess big picture thinking would suggest it would be necessary to adjust for that reason.

zach2good commented on scripts/globals/spells/trust/mumor.lua:
```diff
@@ -2,19 +2,78 @@
 spellObject.onMagicCastingCheck = function(caster, target, spell)
-    return xi.trust.canCast(caster, spell, 1015)
+    return xi.trust.canCast(caster, spell)
 end
```
Can multiple versions of mumor be cast at the same time on retail?

GodModeXI replied:
Mumor and Mumor II

zach2good replied:
Shantotto and Shantotto ii can't be cast at the same time on retail if I remember correctly, is that not the case here?

GodModeXI replied:
Ah, so the spell ID is defined in this case to support the restrictions applied in the trust.lua global. I'll add the spell ID back in.

zach2good commented on scripts/globals/spells/trust/uka_totlihn.lua:
```diff
@@ -5,34 +5,77 @@ require("scripts/globals/trust")
 -----------------------------------
 local spellObject = {}
 
+local jobs1 =
```
`jobs1` could be named something more descriptive, or be named jobs with a comment explaining what it's for

GodModeXI replied:
I will define this better.

GodModeXI committed:
Reviewer requested changes - mumor.lua

resolves healer job table description for more clear use.
moves defined local variables into listener functions to apply directly to mobArg.
adds Todo requests for rework for dynamic synergy.
add Todo request to rework gambit container for toggling behaviors based on synergy.

GodModeXI committed:
sanity check corrections

GodModeXI committed:
sanity check correction

GodModeXI committed:
Reviewer requested changes - uka_totlihn.lua

resolves healer job table description for more clear use.
moves defined local variables into listener functions to apply directly to mobArg.
adds Todo requests for rework for dynamic synergy.
add Todo request to rework gambit container for toggling behaviors based on synergy.

Barthandalus commented:
@zach2good I made updates to the lua files based on your requests. I wrote Todo in on the Gambit container changes and the request for dynamic variable management. a global system may work better. I did this mainly because I don't feel like I have the skillset to work in the core yet. If you think this needs to be closed and resubmitted without synergy intil a better method can be written I can do so on a new PR. Thanks!

GodModeXI committed:
Merge branch 'LandSandBoat:base' into base

zach2good reviewed (commented):
Overall I'm happy with this PR, I think we can get rid of the synergy handling using local vars in favour of using the combat tick listeners:

```lua
    -- For Mumor
    mob:addListener("COMBAT_TICK", "MUMOR_CTICK", function(mobArg)
        local sambaDurationBoost = 0
        local party = mobArg:getMaster():getPartyWithTrusts()
        for _, member in pairs(party) do
            if member:getObjType() == xi.objType.TRUST then
                if
                    member:getTrustID() == xi.magic.spell.UKA_TOTLIHN
                then
                    local sambaDurationBoost = 10
                end
            end
        end
        -- Always set the boost, even if Uka wasn't found.
        -- This accounts for her being in the party and giving the boost
        -- and also if she dies and the boost goes away.
        mobArg:setMod(xi.mod.SAMBA_DURATION, sambaDurationBoost)
    end)

    -- For Uka
    mob:addListener("COMBAT_TICK", "UKA_TOTLIHN_CTICK", function(mobArg)
        local waltzPotencyBoost = 0
        local party = mobArg:getMaster():getPartyWithTrusts()
        for _, member in pairs(party) do
            if member:getObjType() == xi.objType.TRUST then
                if
                    member:getTrustID() == xi.magic.spell.MUMOR or
                    member:getTrustID() == xi.magic.spell.MUMOR_II
                then
                    local waltzPotencyBoost = 10
                end
            end
        end
        -- Always set the boost, even if Mumor wasn't found.
        -- This accounts for her being in the party and giving the boost
        -- and also if she dies and the boost goes away.
        mobArg:setMod(xi.mod.WALTZ_POTENTCY, waltzPotencyBoost)
    end)
```

GodModeXI committed:
Update mumor.lua

Delete variable based synergy handling and replace with a fully contained CTICK check that adjusts the mod value based on member checks in the party

GodModeXI committed:
Update uka_totlihn.lua

removes variable based synergy checks and replaces with a fully contained Ctick check to adjust the value based on member presence

GodModeXI committed:
Update mumor.lua

GodModeXI committed:
Update uka_totlihn.lua

corrected typo

GodModeXI committed:
Merge branch 'LandSandBoat:base' into base
